### PR TITLE
Edit model structure

### PR DESCRIPTION
**Pull request by fvlieshout**
The new model structure is, I think, done like this.
We already adapted train.py, reuters_graph.py and dataset.py so I could at least run the code for the roberta_gnn but we will still need to adapt many things. But that's also very dependent on the datasets structure.
Not sure if we should merge this already or wait for the dataset structures

### CODE REVIEW

File: data_prep/dataset.py [hunk 1]
```diff
@@ -179,6 +183,42 @@ def pre_process_words(self, all_docs):
...
+    def __init__(self, device, n_train_docs, tokenizer):
+        super().__init__(device, n_train_docs)
+        self.tokenizer = tokenizer
+        self.word_embeddings = vocab.GloVe(name='840B', dim=300)
+        self.doc_embeddings = RobertaModel.from_pretrained('roberta-base')
+        self.doc_embeddings.eval()
+    
+    def initialize_data(self, docs, classes):
+        super().initialize_data(docs, classes)
+        self.data.doc_features, self.data.word_features = self.generate_features(self.words)
```
Comment (rahelbeloch) [inline data_prep/dataset.py:196]: If we are anyways calling the `generate_features` only from here, it should probably be a private method (I think that would be `self._generate_features`).

File: data_prep/dataset.py [hunk 2]
```diff
@@ -179,6 +183,42 @@ def pre_process_words(self, all_docs):
...
 
+class RobertaGraphDataset(GraphDataset):
+    def __init__(self, device, n_train_docs, tokenizer):
+        super().__init__(device, n_train_docs)
+        self.tokenizer = tokenizer
+        self.word_embeddings = vocab.GloVe(name='840B', dim=300)
+        self.doc_embeddings = RobertaModel.from_pretrained('roberta-base')
+        self.doc_embeddings.eval()
+    
+    def initialize_data(self, docs, classes):
```
Comment (rahelbeloch) [inline data_prep/dataset.py:194]: From where would this method be called? I think we were planning to just do this automatically once this object is created (so in the init).

File: data_prep/dataset.py [hunk 3]
```diff
@@ -179,6 +183,42 @@ def pre_process_words(self, all_docs):
         """
         raise NotImplementedError
 
+class RobertaGraphDataset(GraphDataset):
```
Comment (rahelbeloch) [inline data_prep/dataset.py:186]: Maybe check how you can already adapt this according to the new skeleton that Matyas created (see https://github.com/fvlieshout/ATCS_group1/pull/19). I think he put this class in a single file already and also made short comments for what should be done where

Comment (rahelbeloch) [reply data_prep/dataset.py:186]: Also, I thiink this class might still miss the preprocess method (https://github.com/fvlieshout/ATCS_group1/blob/b89ba33663e1bfec6559350840ffc2aef424a872/data_prep/reuters_graph.py#L33)

File: data_prep/reuters_graph.py [hunk 4]
```diff
@@ -2,11 +2,11 @@
 
 nltk.download('reuters')
 from nltk.corpus import reuters
-from data_prep.dataset import GraphDataset, Reuters
+from data_prep.dataset import RobertaGraphDataset, Reuters
 
 
-class ReutersGraph(GraphDataset, Reuters):
-    def __init__(self, device, r8=False, val_size=0.1, n_train_docs=None):
+class ReutersGraph(RobertaGraphDataset, Reuters):
```
Comment (rahelbeloch) [inline data_prep/reuters_graph.py:8]: Is this just temporary for now so that the code works still?

File: models/model.py [hunk 5]
```diff
@@ -56,24 +69,35 @@ def training_step(self, batch, batch_idx):
...
         # By default logs it per epoch (weighted average over batches)
-        self.log('val_accuracy', self.accuracy(*self.model(batch, mode='val')))
+        out, labels = self.model(batch, mode='val')
+        predictions = self.classifier(out)
+        self.log('val_accuracy', self.accuracy(predictions, labels))
 
     def test_step(self, batch, batch_idx):
         # By default logs it per epoch (weighted average over batches)
-        self.log('test_accuracy', self.accuracy(*self.model(batch, mode='test')))
+        if self.test_val_mode == 'test':
```
Comment (rahelbeloch) [inline models/model.py]: I thiink you don't even need this if/else here... If you just pass the `mode=self.test_val_mode` to the `self.model`, it will be the correct mode already

File: models/model.py [hunk 6]
```diff
@@ -20,20 +20,33 @@ def __init__(self, model_hparams, optimizer_hparams):
             weight decay, etc.
         """
         super().__init__()
+        # Variable to distinguish between validation and test mask in test_step
+        self.test_val_mode = 'test'
```
Comment (rahelbeloch) [inline models/model.py:25]: (optional) name suggestion: `self.eval_mode` or `self.evaluation_mode`

File: models/model.py [hunk 7]
```diff
@@ -20,20 +20,33 @@ def __init__(self, model_hparams, optimizer_hparams):
...
 
         # Exports the hyperparameters to a YAML file, and create "self.hparams" namespace
         self.save_hyperparameters()
-
         self.loss_module = nn.CrossEntropyLoss()
 
+        roberta_output_dim = 768
```
Comment (rahelbeloch) [inline models/model.py:31]: Do you think it would be sensefull to also put that into the `model_hparams`? Does not have such a high prior for me though and if required could also be done any time later

File: models/model.py [hunk 8]
```diff
@@ -56,24 +69,35 @@ def training_step(self, batch, batch_idx):
             batch_idx     - Index of the batch in the dataset (not needed here).
         """
         # "batch" is the output of the training data loader
-        predictions, labels = self.model(batch, mode='train')
+        out, labels = self.model(batch, mode='train')
+        predictions = self.classifier(out)
```
Comment (rahelbeloch) [inline models/model.py:74]: Could make a method inside the trainer for this, e.g. `def predict(self, batch, mode)` which does these two lines and then this could be called from train step, val step and test step

File: models/model.py (hunk 5 above)
Comment (rahelbeloch) [reply models/model.py]: Oh, although I don't know what Mats plans regarding the transfer are and If that would complicate things for him... So maybe we should consider this optional for now

File: models/model.py [hunk 9]
```diff
@@ -56,24 +69,35 @@ def training_step(self, batch, batch_idx):
...
+        if self.test_val_mode == 'test':
+            out, labels = self.model(batch, mode='test')
+        if self.test_val_mode == 'val':
+            out, labels = self.model(batch, mode='val')
+        predictions = self.classifier(out)
+        self.log('test_accuracy', self.accuracy(predictions, labels))
+    
+    def backward(self, loss, optimizer, optimizer_idx):
+        #override of the backward pass so we can set retain_graph to True
+        loss.backward(retain_graph=True)
```
Comment (rahelbeloch) [inline models/model.py]: Should and is this happening for all models or is it specific to one? (Asking because we hadn't had this before)

Comment (rahelbeloch) [reply models/model.py]: And could we not also do this in the `training_step` as we have the loss there anyways?

File: models/model.py [hunk 10]
```diff
@@ -100,89 +124,76 @@ def get_lr_factor(self, epoch):
...
-        self.conv2 = GCNConv(200, 8)
+class PureGraphEncoder(nn.Module):
+    def __init__(self, input_dim, hidden_dim):
+        super(PureGraphEncoder, self).__init__()
+        self.conv1 = GCNConv(input_dim, hidden_dim)
+        self.conv2 = GCNConv(hidden_dim, hidden_dim)
 
-    def forward(self, data):
+    def forward(self, data, mode):
         x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
```
Comment (rahelbeloch) [inline models/model.py]: I thiink this line is obsolete and can be removed, because `edge_index` and `edge_weight` are anyways overwritten in the next line and `x` is newly initialized in line 162

File: models/model.py [hunk 11]
```diff
@@ -100,89 +124,76 @@ def get_lr_factor(self, epoch):
...
+        edge_index, edge_weight = data.edge_index, data.edge_attr
+        doc_feats = data.doc_features
+        word_feats = data.word_features
+        word_feats = self.linlay(word_feats)
```
Comment (rahelbeloch) [inline models/model.py]: does this class have an attribute `self.linlay`? I don't see it

File: models/model.py [hunk 12]
```diff
@@ -100,89 +124,76 @@ def get_lr_factor(self, epoch):
...
-        self.model = GraphNet(model_hparams['num_nodes'])
-        self.test_val_mode = 'test'
+class RobertaGraphEncoder(nn.Module):
+    def __init__(self, input_dim, hidden_dim):
+        super(RobertaGraphEncoder, self).__init__()
+        self.linlay = nn.Linear(300, 768)
+        self.conv1 = GCNConv(input_dim, hidden_dim)
+        self.conv2 = GCNConv(hidden_dim, hidden_dim)
 
     def forward(self, data, mode):
```
Comment (rahelbeloch) [inline models/model.py]: Ok, now I am a bit confused... there is exactly the same stuff happening in `RobertaGraphEncoder` and `PureGraphEncoder` forward methods ... is this part still pending or is this supposed to be like that? 
Because If it is supposed to be so, then why do we need two different model classes for this? Do I understand correctly then, that the only difference between Roberta Graph Encoder and Pure Graph Encoder is the way the data is prepared beforehand?

File: train.py [hunk 13]
```diff
@@ -48,10 +48,12 @@ def train(model_name, seed, epochs, patience, b_size, l_rate, w_decay, warmup, m
     model_params = {
         'model': model_name,
         'cf_hid_dim': cf_hidden_dim,
+        'gnn_output_dim': additional_params.get('gnn_output_dim'),
+        'num_classes': additional_params['num_classes'],
```
Comment (rahelbeloch) [inline train.py]: These two lines are not needed, that is exactly what the `**additional_params` in the next line does, it adds all entries from `additional_params` to the `model_params` dict

**Review by rahelbeloch — CHANGES_REQUESTED** [review]
@fvlieshout Nice, looks good. I added some comments... 

I think once you aligned this as much as possible with the new skeleton suggested by Matyas in his PR, I would opt to not wait with merging until the other things are done. We anyways will have to to it incrementally and I think this is a good basis to start with

File: data_prep/dataset.py [hunk 14]
```diff
@@ -7,6 +7,8 @@
 from data_prep.graph_utils import tf_idf_mtx, get_PMI
```
Comment (Matyasch) [inline data_prep/dataset.py:7]: So I assume you know that this file will be rewritten and you just used it for testing. I will use it for inspiration tho

File: data_prep/dataset.py (hunk 3 above)
Comment (Matyasch) [reply data_prep/dataset.py:186]: These implementation details are "hidden" from the models anyways so I think it doesn't really matter how and in what files these classes are right now. We will do the datasets in a way that they have the same interface as what the models expect. So in the end only the import paths have to be changed and maybe some small things in train.py

File: data_prep/reuters_graph.py [hunk 15]
```diff
@@ -2,11 +2,11 @@
```
Comment (Matyasch) [inline data_prep/reuters_graph.py:2]: Same as dataset.py except this will be completely gone

File: models/model.py (hunk 9 above)
Comment (Matyasch) [inline models/model.py]: Did you find out why this is needed?

File: models/model.py [hunk 16]
```diff
@@ -100,89 +124,76 @@ def get_lr_factor(self, epoch):
...
-        inputs, attention_mask = batch['input_ids'], batch['attention_mask']
-
-        out = self.encoder(inputs, attention_mask)
-        out = self.classifier(out)
-        return out, batch['labels']
-
-
-class TransformerModel(nn.Module):
-
+class RobertaEncoder(nn.Module):
```
Comment (Matyasch) [inline models/model.py]: This class should go into its own file `models/roberta_encoder.py` to avoid multi-purpose and long files. Let's use that `models` directory

File: models/model.py [hunk 17]
```diff
@@ -100,89 +124,76 @@ def get_lr_factor(self, epoch):
...
         # returns a tuple of torch.FloatTensor comprising various elements depending on the (RobertaConfig) and inputs.
         hidden_states = self.model(inputs, attention_mask)
 
         # b_size x hid_size
-        cls_token_state = hidden_states[1]
-
-        return cls_token_state
-
+        out = hidden_states[1]
+        return out, batch['labels']
```
Comment (Matyasch) [inline models/model.py]: So I assume we need to return the labels from the dataset in the model because only the model knows how its dataset works. Interesting

File: models/model.py (hunk 11 above)
Comment (Matyasch) [reply models/model.py]: Good catch, this is an incorrect copy from the robertaGNN model. We don't have `doc_feats` or `word_feats`, we just get `x` immediately.

File: models/model.py [hunk 18]
```diff
@@ -100,89 +124,76 @@ def get_lr_factor(self, epoch):
...
 
-    def forward(self, data):
+    def forward(self, data, mode):
         x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
+        edge_index, edge_weight = data.edge_index, data.edge_attr
+        doc_feats = data.doc_features
+        word_feats = data.word_features
+        word_feats = self.linlay(word_feats)
+        x = torch.cat((doc_feats, word_feats))
         x = self.conv1(x, edge_index, edge_weight)
```
Comment (Matyasch) [inline models/model.py]: So the actual second line of this function is this, everything in between should be deleted

File: train.py [hunk 19]
```diff
@@ -173,7 +181,8 @@ def get_dataset(dataset_name):
     elif dataset_name == "AGNewsText":
         return AGNewsText
     elif dataset_name == 'R8Graph':
-        return R8Graph(device)
+        tokenizer = RobertaTokenizerFast.from_pretrained("roberta-base")
```
Comment (Matyasch) [inline train.py:177]: Naturally, this function will have to be rewritten but that's not a concern right now

**Review by Matyasch — CHANGES_REQUESTED** [review]
I think it would be nice to have all models in their individual files to avoid cluttered files with many purposes. You should change the PureGNN model and otherwise this looks good

File: data_prep/dataset.py (hunk 14 above)
Comment (fvlieshout) [reply data_prep/dataset.py:7]: Exactly

File: data_prep/dataset.py (hunk 3 above)
Comment (fvlieshout) [reply data_prep/dataset.py:186]: Yes this file I just adapted for testing. It needs to be changed into the new dataset structure

File: data_prep/dataset.py (hunk 2 above)
Comment (fvlieshout) [reply data_prep/dataset.py:194]: Yes that needs to be implemented but I only put it here now for testing. This needs to be updated by Matyas

File: data_prep/dataset.py (hunk 1 above)
Comment (fvlieshout) [reply data_prep/dataset.py:196]: Agreed. Maybe @Matyasch can take care of that when he implements the dataset structures

File: data_prep/reuters_graph.py (hunk 15 above)
Comment (fvlieshout) [reply data_prep/reuters_graph.py:2]: Indeed

File: data_prep/reuters_graph.py (hunk 4 above)
Comment (fvlieshout) [reply data_prep/reuters_graph.py:8]: yes

File: models/model.py (hunk 6 above)
Comment (fvlieshout) [reply models/model.py:25]: I think it's clearer like this because it differentiates between test and validation mode

File: models/model.py (hunk 7 above)
Comment (fvlieshout) [reply models/model.py:31]: I thought about that as well but because it will definitely always stay the same (we cannot really change Roberta), I opted to let it be defined here

File: models/model.py (hunk 8 above)
Comment (fvlieshout) [reply models/model.py:74]: I don't really agree. It will only be two lines and making a separate method for that would be less clear in my opinion

File: models/model.py (hunk 5 above)
Comment (fvlieshout) [reply models/model.py]: No you are right. I changed it. This shouldn't make a difference for the implementation of Mat

File: models/model.py (hunk 9 above)
Comment (fvlieshout) [reply models/model.py]: Yes. Because in the RobertaGraph, the input is a concatenatation of the document features and the linear projection of the word features. This means that it has a slightly more complicated 'graph' to backpropagate to as it needs to remember that the upper rows of the feature matix came from Roberta (which it doesn't need to update) and the lower rows of the feature matrix came from the linear projection of the word features (which it does need to update). Usualy Pytorch throws away this information to save memory but in this case, it can't throw that information away and we have to let that know by setting retain_graph=True.
This is only used for the RobertaGraph model btw but it doesn't work if I just overwrite the backward method of the RobertaGraph encoder. If this results in memory issues for the other encoders, I can take another look at if we can only do this for the RobertaGraphEncoder.

Comment (fvlieshout) [reply models/model.py]: No I have to specifically overwrite the backward method. See the reply to Matyas' comment

File: models/model.py (hunk 10 above)
Comment (fvlieshout) [reply models/model.py]: Yes good catch. I removed it

File: models/model.py (hunk 12 above)
Comment (fvlieshout) [reply models/model.py]: Yes I wrongly copied some stuff in the PureGraphEncoder that actually should only have been in the RobertaGraphEncoder. I fixed it now.
The main difference is that the features are obtained differently

File: models/model.py (hunk 9 above)
Comment (rahelbeloch) [reply models/model.py]: I would prefer to only do this for the RobertaGraph Encoder. can you not simply wrap this with a `iif(isinstance(self.model, RobertaGraphEncoder)){ ... }`?

File: data_prep/dataset.py (hunk 1 above)
Comment (Matyasch) [reply data_prep/dataset.py:196]: oki

**Review by Matyasch — APPROVED** [review]

File: models/model.py (hunk 12 above)
Comment (rahelbeloch) [reply models/model.py]: Ok, well but then I think the `RobertaGraphEncoder` should actually only contain this:

Screenshot 2021-05-15 at 19 12 42

But it's ok for now I guess

**Review by rahelbeloch — APPROVED** [review]
You can merge from my side
